File: 0035-search-insert-position/0035-search-insert-position.py
class Solution:
    def searchInsert(self, nums: List[int], target: int) -> int:
        '''
        0, 1, 2, 3
        1, 3, 5, 6
        l  m     r
        lr
        r  l=res

        0, 1, 2, 3
        1, 3, 5, 6  res = 0

        '''

        l, r = 0, len(nums) - 1

        while l <= r:
            m = l + (r - l) // 2
            if nums[m] == target:
                return m
            elif nums[m] < target:
                l = m + 1
            else:
                r = m - 1
        return l

File: 0035-search-insert-position/0035-search-insert-position.cpp
class Solution {
public:
    int searchInsert(vector<int>& nums, int target) {
        int left = 0, right = nums.size() - 1;
        while (left <= right) {
            int mid = left + (right - left) / 2;
            if (nums[mid] == target) 
                return mid;
            else if (nums[mid] < target) {
                left = mid + 1;
            }
            else {
                right = mid - 1;
            }
        }
        return left;
    }   
};

/*
0,1,2,3
1,3,5,6
l.m   r
lr
r l
return l

1,3,5,6
l m.  r
    lmr
      lmr
      r l
*/